### DISCUSSION TITLE
Detect when `py` is part of a symlink loop

### DISCUSSION THREAD

**Discussion by cactusdualcore** [brettcannon/python-launcher#173] [Ideas]
**Launcher will use itself as python if python is a symlink to py**
If the installation of python found by `py` to be the best-suited installation is a symbolic link to `py` itself, it will call itself again, resulting in a unending loop.

**To Reproduce**
Steps to reproduce the behaviour:
1. Install [py](https://crates.io/crates/python-launcher)
2. Install [venv-wrapper](https://crates.io/crates/venv-wrapper)
3. Add `py` to `$PATH` so `$ which py` shows installation of py
4. Add `venv-wrapper` to `$PATH`
5. create a virtual environment with `$ venv-wrapper new <name> -p "$(which py)"`
6. activate virtual environment with `$ venv-wrapper use <name>`
7. execute `$ python --version`. No output will appear, because `py` calls itself repeatedly.

**Expected behaviour**
`py` should not call itself but either error or look for an actual python install.

**Screenshots**
![a picture of my terminal](https://user-images.githubusercontent.com/64769435/156047157-dc02fd25-b273-4cd5-a446-0e9974ad3fc9.png)
`$ python --version` blocks until interrupted (e.g. by CTRL + C)

**System Details (please complete the following information):**
 - OS: Linux Mint 20.3 x86_64
 - Shell: zsh 5.8
 - Launcher Version: 1.0.0

**Comment by brettcannon** [comment]
Why was the Python Launcher symlinked to `python`? What is `venv-wrapper new <name> -p "$(which py)"` expected to do?

**Comment by brettcannon** [comment]
If you're trying to get the path to the executable that the Launcher chooses, you could run `py -c "import sys; print(sys.executable)"`.

**Comment by cactusdualcore** [comment]
> Why was the Python Launcher symlinked to `python`?

Curiosity made me try whether `py` is a replacement to `python` and `python3` to where it can be used as the referenced executable of a virtual python environment.
> What is `venv-wrapper new <name> -p "$(which py)"` expected to do?

My assumption was for `py` to notice it is calling itself, warn the user this is unexpected and, if full interchangeability between `py` and `python` is wanted, to fall back to an actual python installation. I understand the latter might not be what `py` is trying to do, though I think a warning may be helpful if this arose by misunderstanding or similar.

**Comment by cactusdualcore** [comment]
> If you're trying to get the path to the executable that the Launcher chooses, you could run `py -c "import sys; print(sys.executable)"`.

This should work, though it might not be very useful. One big advantage of virtual python environments is to provide the same version of python for an application to be run with. If a new global install of python was considered more fit by the Launcher the edition of python could change.

**Comment by brettcannon** [comment]
> I understand the latter might not be what `py` is trying to do, though I think a warning may be helpful if this arose by misunderstanding or similar.

How would the Python Launcher detect this?

> One big advantage of virtual python environments is to provide the same version of python for an application to be run with.

Correct, but your example is calling something `venv-wrapper` which suggests you're trying to create a virtual environment. The line I gave you was for potential use as the argument to that `-p` option.

BTW, I consider this a feature request, so I am going to convert it to a Discussion.

**Reply by cactusdualcore** [reply]
A possible solution to the latter problem might be to compare the [executable path](https://doc.rust-lang.org/std/env/fn.current_exe.html) and the found binary.

**Reply by cactusdualcore** [reply]
> **S_ISLNK(m)**
>
> symbolic link? (Not in POSIX.1-1996.)

It seems POSIX does not require an API to deal with symlinks and it seems as nix does not support checking whether a path is. Though [Microsoft Windows](https://docs.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants), [Linux](https://linux.die.net/man/2/fstat) and [MacOS](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fstat.2.html) all have support for them.

**Reply by cactusdualcore** [reply]
The suggested check's results may change for different platforms. Especially on systems which allow hardlinks this can fail.
